### PR TITLE
[go_router] Exposes package-level privates

### PR DESCRIPTION
**Pull request by chunhtai**
## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/delegate.dart [hunk 1]
```diff
@@ -26,7 +25,6 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     required GoRouterWidgetBuilder? errorBuilder,
     required List<NavigatorObserver> observers,
     required this.routerNeglect,
-    required this.onException,
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:29]: This was a mistake I made earlier. this property is unused

**Review by johnpryan — APPROVED** [review]

File: packages/go_router/lib/src/router.dart [hunk 2]
```diff
@@ -295,9 +295,6 @@ class GoRouter implements RouterConfig<RouteMatchList> {
   @override
   late final GoRouteInformationParser routeInformationParser;
 
-  /// Returns `true` if there is at least two or more route can be pop.
-  bool canPop() => routerDelegate.canPop();
```
Comment (ahmednfwela) [inline packages/go_router/lib/src/router.dart:299]: why was this removed while pop is still there?

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:299]: you are right, I keep thinking this as a state that should be remove, but it is actually needed for pop as well.

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart]: added it back and updated migration guide

Comment (hrishikesh-kadam) [reply packages/go_router/lib/src/router.dart]: @chunhtai Also consider correcting CHANGELOG.md for v9.0.0

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart]: done
